base/io.jl

```diff
@@ -445,39 +443,80 @@ function readuntil{T}(s::IO, delim::T)
     return out
 end
 
-# based on code by Glen Hertz
-function readuntil(s::IO, t::AbstractString)
-    l = length(t)
-    if l == 0
+function _rfind_sub_c(target, endc, endstate)
+    # find the largest i (< endstate) such that "target[(end - i + 1):(end - 1)] * c" == "target[1:i]"
+    # (equivalently, find the smallest tii' such that "target[tii':(end - 1)] * c" == "target[1:(end - tii' + 1)]")
+    # this assumes that the `target` iterator is pure and supports equality comparison for its state
+    # and that endstate is a valid state for target
+    tii = start(target)
+    i = start(target)
+    ti = tii
```

Comment:
I wrote it this way because it mirrors the ordering needed to update the iterator in the while loop